crates/pipeline-manager/src/runner/pipeline_executor.rs

```diff
@@ -0,0 +1,251 @@
+                            // As such, this case is not erroneous.
+                            debug!("Log follow request sender itself closed, which happens when the pipeline is deleted. Log rejection thread is ended for this reason.");
+                            break;
+                        }
+                    }
+                }
+            }
+            log_follow_request_receiver
+        });
+        debug!(
+            "Logging request are being rejected for pipeline {}",
+            pipeline_id
+        );
+        (terminate_sender, join_handle)
+    }
+
+    /// Terminates the log thread (either rejection of operational) by sending the termination
+    /// message and joining the thread.
+    /// Returns the log follow request receiver.
+    async fn terminate_log_thread(
+        terminate_sender: &mut Sender<()>,
+        join_handle: &mut JoinHandle<Receiver<Sender<LogMessage>>>,
+    ) -> Receiver<Sender<LogMessage>> {
+        // The terminate receiver might have been dropped already if there was an unrecoverable
+        // error. If this occurred, the thread will have exited by itself. As such, it is not
+        // needed to check whether send was a success or not.
+        let _ = terminate_sender.send(()).await;
+        join_handle
+            .await
+            .expect("Unable to receive joined log thread result")
+    }
+
+    /// Catches up the follower by sending all the buffered logs to it.
+    /// Afterward, adds it to the list of known followers if there was
+    /// no error during sending the catch-up.
+    async fn catch_up_and_add_follower(
+        logs: &mut LogsBuffer,
+        log_followers: &mut Vec<Sender<LogMessage>>,
+        follower: Sender<LogMessage>,
+    ) {
+        // Catch up the new follower if there is any lines to catch-up on
+        let mut failed = false;
+        if logs.num_discarded_lines() == 0 || follower.send(LogMessage::Line(format!("... {} prior log lines were discarded due to buffer constraints and are thus not shown.", logs.num_discarded_lines()))).await.is_ok() {
+            for line in logs.lines() {
+                if follower.send(LogMessage::Line(line.clone())).await.is_err() {
+                    failed = true;
+                    break;
+                }
+            }
+        } else {
+            failed = true;
+        }
+
+        // Any failure in sending results in the follower Sender not being added
+        // to the list, and thus going out of scope and being dropped.
+        // The Receiver in that case will be notified no Sender exists anymore.
+        if !failed {
+            log_followers.push(follower);
+        }
+    }
+
+    /// Process a new log line by adding it to the lines buffer and
+    /// sending it out to all followers. Any followers that exhibit
+    /// a send error are removed.
+    async fn process_log_line_with_followers(
```

Comment:
Can this stuff be simplified by using https://docs.rs/tokio/1.40.0/tokio/sync/broadcast/index.html

Reply:
Unfortunately not I think, at the start we need to catch up a new follower and need to send to only that channel and not all the other existing followers.